accord-core/src/main/java/accord/local/SafeCommand.java

```diff
@@ -131,14 +143,27 @@ public Command.Executed preapplied(CommonAttributes attrs, Timestamp executeAt,
         return update(Command.preapplied(current(), attrs, executeAt, waitingOn, writes, result));
     }
 
+    public Command.Executed applying()
+    {
+        return update(Command.applying(current().asExecuted()));
+    }
+
     public Command.Executed applied()
     {
         return update(Command.applied(current().asExecuted()));
     }
 
-    public Command.NotWitnessed notWitnessed()
+    public Command.NotDefined uninitialised()
     {
         Invariants.checkArgument(current() == null);
-        return update(Command.NotWitnessed.notWitnessed(txnId));
+        return update(Command.NotDefined.uninitialised(txnId));
+    }
+
+    public Command initialise()
+    {
+        Command current = current();
+        if (current.saveStatus() != Uninitialised)
+            return current;
+        return update(Command.NotDefined.notDefined(current, current.promised()));
```

Comment:
`saveStatus == Uninitialised`, so wouldn't this return `NotDefined` with a `null` for promised (if we saw an invalidate we transitioned to `NotDefined`; so wouldn't that mean this returns a `NotDefined` with the same semantics as `Uninitialised`?

Reply:
the caller is

```
             // ensure we have a record to work with later; otherwise may think has been truncated
            blockedBy.initialise();
```

so I guess we do want `NotDefined(promised = null)`, so the distinction is getting blurred to me... so if we depend on a txn, or saw it as invalidated before we saw its definition, then we are `NotDefined`... but how did we learn about this txn id if we didn't learn it from a peer?

Reply:
Looking at `accord.local.SafeCommandStore#get(accord.primitives.TxnId, accord.primitives.Unseekables<?>)` I guess we can do the following...


```
Applied -> Truncated -> Uninitialised // caused by truncate
```

And ``accord.impl.AbstractSafeCommandStore#get says that we asked for a txn, but there was no known txn, so we return `Uninitialised`...  So this happens when we see an unknow txn for the first time, or it was purged... but NotDefined is for txn we learn from peers...  which feels confusing.

Reply:
Uninitialised is a a local state that just means we haven’t initialised the transaction state yet. NotDefined is set as soon as we actually record anything associated with the transaction. 

We can go back to NotWitnessed, or perhaps there’s a third better descriptor.

Note though that we adopt this state after BeginInvalidation also, if we haven’t previously been PreAccepted. So NotWitnessed feels more inaccurate to me.

We could always introduce a PreAcceptWithoutDefinitiom status to delineate that state though.